### PR TITLE
Pytorch integration for edgeml

### PR DESCRIPTION
**Pull request by adityakusupati**
- [X] Bonsai - Fully functional on CPU and GPU along with SeeDot support
- [X] FastCells - Done - Fully functional on CPU and GPU
- [x] Figure out a package structure for pytorch and tf
- [x] ProtoNN porting to pytorch with CPU support
- [x] ProtoNN porting to pytorch with GPU support
- [x] SRNN port to pytorch
- [x] SRNN examples with Google-13
- [x] SRNN for GPU
- [x] Supporting dropout parameters for SRNN
- [x] Add `argparse` support for SRNN example

@metastableB and @harsha-simhadri I am not sure how to disambiguate both tf and pytorch packages hence naming this pytorch_edgeml for now. This can be changed.

Note: @kunaldahiya and I tried to get hardThrsd function completely onto GPU, but failed due to issues in PyTorch. Gradient flow isn't straightforward for hardThrsd() or any function if the numpy and torch share the same memory. Hence have to replicate.

### CODE REVIEW

**Comment by metastableB** [comment]
@adityakusupati  I do not understand how pytorch works yet. You can take a call on this PR once Harsha's review is done.

As far as  packaging is concerned, we should ideally have two packages `tfedgeml` and a `pytedgeml`.  For now your way of disambiguation works.

**Comment by adityakusupati** [comment]
@metastableB sure I will wait for @harsha-simhadri's review.

I would prefer something else for pytedgeml, I will leave the way it is right now, but tis does need fixing so that it doesn't confuse/conflict fro tf/pytorch.

File: pytorch/pytorch_edgeml/graph/bonsai.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.Z = self.initZ(Z)
+
+        self.assertInit()
+
+    def initZ(self, Z):
+        if Z is None:
+            Z = torch.randn([self.projectionDimension, self.dataDimension])
+            Z = nn.Parameter(Z)
+        else:
+            Z.data = torch.from_numpy(Z.astype(np.float32))
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/bonsai.py]: Is this a bug? Type of `Z` is unclear here. How can you do `Z.data` ?

File: pytorch/pytorch_edgeml/trainer/bonsaiTrainer.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        self.lV = lV
+        self.lT = lT
+        self.lZ = lZ
+
+        self.sW = sW
+        self.sV = sV
+        self.sT = sT
+        self.sZ = sZ
+
+        if device is None:
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/trainer/bonsaiTrainer.py]: You can probably move this to `train` routine ? That makes more sense ? So that user can switch between `cpu` and `gpu` over different calls to `train` ? 

Also will help with bench-marking i think and debugging.

Comment (metastableB) [reply pytorch/pytorch_edgeml/trainer/bonsaiTrainer.py]: In fact if you think about it, because of the way pytorch works, all the arguments provided to `__init__` can be moved to `train` method and can be dynamically changed.

File: pytorch/pytorch_edgeml/utils.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    dest = dest.reshape(-1)
+    dest[zeroSupport] = 0
+    dest = dest.reshape(src.shape)
+    del zeroSupport
+    return dest
+
+
+def countnnZ(A, s, bytesPerVar=4):
+    '''
+    Returns # of non-zeros and representative size of the tensor
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/utils.py]: Please explain arguments A and s ?

File: pytorch/pytorch_edgeml/utils.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    '''
+    zeroSupport = (src.view(-1) == 0.0).nonzero()
+    dest = dest.reshape(-1)
+    dest[zeroSupport] = 0
+    dest = dest.reshape(src.shape)
+    del zeroSupport
+    return dest
+
+
+def countnnZ(A, s, bytesPerVar=4):
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/utils.py]: Plz document

File: README.md [hunk 5]
```diff
@@ -56,6 +57,19 @@ us](mailto:edgeml@microsoft.com).
 to this
 [project](https://www.microsoft.com/en-us/research/project/resource-efficient-ml-for-the-edge-and-endpoint-iot-devices/).
 
+If you use the EdgeML library in your projects or publications, please use the
+following BibTeX to cite us. Note: Author ordering is based on last names.
```
Comment (adityakusupati) [inline README.md]: @harsha-simhadri , this will be completely your call. As of now, we have the repo citation to be last name basis. You make the final decision on this citation. We feel that repos should have bibtex entries as Extreme Classification repo has. I have seen at least 10 papers using EdgeML codebases.

**Comment by metastableB** [comment]
@harsha-simhadri @adityakusupati  I'm done on my part. You can proceed to merge.

**Comment by harsha-simhadri** [comment]
pytorch yet to be tested thoroughly.
SRNN2 dropout fix in later commit. Will add in new PR.
